src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs

```diff
@@ -12194,6 +12193,26 @@ public void TestAnalyzerFilteringBasedOnSeverity(DiagnosticSeverity defaultSever
             }
         }
 
+        [Fact]
+        public void SourceGenerators_EmbeddedSources()
+        {
+            var dir = Temp.CreateDirectory();
+            var src = dir.CreateFile("temp.cs").WriteAllText(@"
+class C
+{
+}");
+
+            var generatedSource = "public class D { }";
+            var generator = new SimpleGenerator(generatedSource);
+
+            VerifyOutput(dir, src, includeCurrentAssemblyAsAnalyzerReference: false, additionalFlags: new[] { "/langversion:preview", "/debug:embedded", "/out:embed.exe" }, generators: new[] { generator }, analyzers: null);
+
+            ValidateEmbeddedSources_Portable(new Dictionary<string, string> { { Path.Combine(dir.Path, SimpleGenerator.FileName), generatedSource } }, dir, true);
+
+            // Clean up temp files
+            CleanupAllGeneratedFiles(src.Path);
+        }
```

Comment:
Need to test some more cases:
- Multiple generated files
- Multiple generated files with identical hint values
- Case where an empty file is returned


Generally do we have tests where the hint path contains known invalid file name characters?

Reply:
Added more tests, but ran into #42628, which I've now addresses as part of this PR too.

We aren't currently putting these down on disk, just using them as identifiers in the PDB, so we should be good from a characters perspective. We'll need to address it when we put them to disk though (presumably some form of encoding?)

---
In reply to: [434220765](https://github.com/dotnet/roslyn/pull/44768#discussion_r434220765) [](ancestors = 434220765)